pyro/poutine/lower_poutine.py

```diff
@@ -0,0 +1,36 @@
+from __future__ import absolute_import, division, print_function
+
+from torch.autograd import Variable
+import pyro  # XXX remove this cyclic dependency
+from pyro.params import _PYRO_PARAM_STORE
+
+from .poutine import Poutine
+
+
+class LowerPoutine(Poutine):
+    """
+    Poutine which "lowers" sample sites to param + observe sites.
+
+    Given a stochastic function with sample calls from a prior, creates a
+    stochastic function where each sample site is replaced by a param site
+    (denoting the parameter) plus an observe site (denoting the prior).
+
+    :param fn: stochastic function
+    """
+    def _pyro_sample(self, msg):
+        if msg["is_observed"]:
+            return super(LowerPoutine, self)._pyro_sample(msg)
+        param_name = msg["name"]
+        msg["name"] = "prior.{}".format(param_name)
+        if param_name in _PYRO_PARAM_STORE:
+            # Treat existing param value as an observation.
+            msg["is_observed"] = True
+            msg["value"] = _PYRO_PARAM_STORE.get_param(param_name)
+            value = super(LowerPoutine, self)._pyro_sample(msg)
+        else:
+            # Sample an initial param value from the prior.
+            value = super(LowerPoutine, self)._pyro_sample(msg)
+            value = Variable(value.data, requires_grad=True)
+            msg["value"] = value
+            msg["is_observed"] = True
+        return pyro.param(param_name, value)
```

Comment:
Wouldn't it be better if `param_name` were changed and `msg["name"]` were preserved?

Reply:
I tried that at first, but it seemed awkward. We have to decide between pairs of site names, either {"mu", "prior.mu"} or {"mu", "lowered.mu"}. I figure that since the user will almost always want to read the `param` site and almost never want to read the `observe` site, it would be better to preserve the param site. This also has the benefit that the site name will be stored in the param store, which is the same effect that `pyro.param` has (i.e. the statement label ends up in the param store).

Reply:
>  I figure that since the user will almost always want to read the param site and almost never want to read the observe site, it would be better to preserve the param site.

How should users be getting the results of inference?  In SVI, they can get stuff from the guide, and they can always use `Marginal`, but for `MAP` they're supposed to query the parameter store?

Reply:
As I understand, the results of inference are optimized parameters. For both SVI and MAP, users get those results from the parameter store.

Reply:
This also has the property that `pyro.lower(pyro.lift(fn, prior))` merely inserts `pyro.observe("prior.*", ...)` sites for each parameter (i.e. the param sites are preserved).